File: python/segTree.py
"""
    Segment Tree
"""
import math

class segTree:

    def __init__(self, n, op, initval):
        self.size = 2**(math.ceil(math.log(n, 2)) + 1)
        self.op = op
        self.init_val = initval

        self.v = [self.init_val]*self.size

    def fill(self, p):
        if p >= self.size//2:
            return self.v[p]
        
        self.v[p] = self.op(self.fill(2*p), self.fill(2*p+1))
        return self.v[p]

    def cal(self, p, L, R, nodeL, nodeR):
        if R < nodeL or nodeR < L:
            return self.init_val
        if L <= nodeL and nodeR <= R:
            return self.v[p]

        m = (nodeL + nodeR)//2
        return self.op(self.cal(2*p, L, R, nodeL, m), self.cal(2*p + 1, L, R, m + 1, nodeR))

    def update(self, p, v):
        p += self.size//2 - 1
        self.v[p] = v

        while p > 1:
            p //= 2
            self.v[p] = self.op(self.v[2*p], self.v[2*p + 1])

File: c++/segTree.cpp
/*
    Segment Tree
*/

#include "segTree.h"
#include <algorithm>

namespace boj
{

    void segTree::init(int n, int v)
    {
        while (n*2 > size) size *= 2;
        init_val = v;
        std::fill(a, a+size, init_val);
    }

    int segTree::fill(int p)
    {
        if (p >= size/2) return a[p];

        a[p] = fill(2*p) + fill(2*p+1);
        return a[p];
    }

    int segTree::cal(int p, int L, int R, int nodeL, int nodeR)
    {
        if (R < nodeL || nodeR < L)
            return init_val;
        if (L <= nodeL && nodeR <= R)
            return a[p];

        int m = (nodeL + nodeR)/2;
        return cal(2*p, L, R, nodeL, m) + cal(2*p + 1, L, R, m + 1, nodeR);
    }

    void segTree::update(int p, int v)
    {
        p += size/2 - 1;
        a[p] = v;

        while (p > 1)
        {
            p /= 2;
            a[p] = a[2*p] + a[2*p + 1];
        }
    }

}
